src/nupic/research/connections.py

```diff
@@ -52,92 +162,181 @@ def __init__(self,
-    self._nextSegmentIdx = 0
-    # Index of the next synapse to be created
-    self._nextSynapseIdx = 0
+    self._numSegments = 0
+    self._numSynapses = 0
+    self._nextFlatIdx = 0
+    self._iteration = 0
 
 
-  def cellForSegment(self, segment):
-    """ Returns the cell that a segment belongs to.
+  def segmentsForCell(self, cell):
+    """ Returns the segments that belong to a cell.
 
-    @param segment (int) Segment index
+    @param cell (int) Cell index
 
-    @return (int) Cell index
+    @return (list) Segment objects representing segments on the given cell
     """
-    return self._segments[segment]
 
+    segmentsList = self._cells[cell].segments
 
-  def columnForSegment(self, segment, cellsPerColumn):
-    """ Returns the column that a segment's presynapticCell belongs to
 
-    @param segment        (int) Segment index
-    @param cellsPerColumn (int) Number of cells in a column in the tm
-
-    @return (int)
-    """
-    return self._segments[segment] / cellsPerColumn
+    return [Segment(i, cell, segmentsList[i])
+            for i in xrange(len(segmentsList))
+            if not segmentsList[i].destroyed]
 
 
-  def segmentsForCell(self, cell):
-    """ Returns the segments that belong to a cell.
+  def synapsesForSegment(self, segment):
+    """ Returns the synapses on a segment.
 
-    @param cell (int) Cell index
+    @param segment (int) Segment index
 
-    @return (set) Segment indices
+    @return (list) Synapse objects representing synapses on the given segment
     """
-    self._validateCell(cell)
 
-    if not cell in self._segmentsForCell:
-      return set()
+    segmentData = segment.data
+    if segmentData.destroyed:
+      raise ValueError("Attempting to access destroyed segment's synapses")
 
-    return self._segmentsForCell[cell]
+    return [synapse
+            for synapse in self._synapsesForSegment[segmentData.flatIdx]
+            if not synapse.data.destroyed]
 
 
   def dataForSynapse(self, synapse):
     """ Returns the data for a synapse.
 
-    @param synapse (int) Synapse index
+    @param synapse (Object) Synapse object
 
     @return (SynapseData) Synapse data
     """
 
-    return self._synapses[synapse]
+    return synapse.data
 
 
-  def synapsesForSegment(self, segment):
-    """ Returns the synapses on a segment.
+  def dataForSegment(self, segment):
+    """ Returns the data for a segment.
 
-    @param segment (int) Segment index
+    @param segment (Object) Segment object
+
+    @return (SegmentData) Segment data
+    """
+
+    return segment.data
+
+
+  def retrieveSegmentData(self, idx, cell):
+    """ Returns the data for a segment stored in the self._cells array
+
+    @param idx  (int) segment index on a cell
+    @param cell (int) cell index
+
+    @return (SegmentData) Segment data
 
-    @return (set) Synapse indices
+    NOTE: Only used in the read method in temporal memory for re-creating the
+          active and matching segments
```

Comment:
Instead, maybe there should be a `connections.getSegment` or `connections.segment` method which takes a cell and segment index and returns a `Segment` object. Then `temporal_memory` won't have to instantiate Segments.